### ISSUE TITLE
Rolling upgrades with local volumes can lead to unschedulable Pods

### ISSUE THREAD

**Issue by sebgl** [elastic/cloud-on-k8s#3088]
Any nodeSet modification (podTemplate, Elasticsearch config, etc.) normally leads to a rolling upgrade of the impacted Pods. The idea here is to replace the Pod but reuse the data volume, which is somehow equivalent to "restarting" the Elasticsearch node.

When using local PersistentVolumes, the replacing Pod can only be scheduled on the same Kubernetes node, usually through an affinity constraint in the PersitentVolume resource.

This can cause a rolling upgrade to stay stuck. Examples:

* The user requested a nodeSet memory requirements increase from 8GB to 16GB, but the node cannot hold a 16GB Pod. The 16GB Pod will stay Pending forever.
* Between the Pod deletion (done by ECK) and re-creation (done by the StatefulSet controller), another Pod gets scheduled on the Kubernetes node, making use of the empty allocation slot. The replacing Pod cannot be scheduled anymore and will stay Pending forever.

Even though the Pod could technically be scheduled on another Node, we're stuck here.

When that situation happens, there are "manual" workarounds:

* Delete other Pods on the Node so our Pending Pod can be scheduled.
* Revert the requirements change (eg. 8GB to 16GB memory bump) so the same Node can hold the Pod again.
* Delete the PersistentVolume and PersistentVolumeClaim and the Pending Pod, so a new empty Volume can be assigned to the Pod, that may now be scheduled on another Node.
* Copy/Move the local PersistentVolume to another Node.

Currently ECK makes no assumptions about whether a Pod can be scheduled, and does not even know about Kubernetes Nodes. It also does not care about PersistentVolume affinity constraints.

We may want to explore solutions to better accommodate this. The right solution may be to just do nothing though, since this may be solved in K8s eventually.

Some ideas:

* This is a generic problem to solve in K8s, not something to necessarily solve in ECK.
* Improvement in local volumes driver and provisioner could eventually account for this situation and handle it automatically.
* The user could indicate they want PVs and Pods to be deleted automatically if a Pod stays Pending for X seconds/minutes. It would force the Pod to be rescheduled elsewhere with a fresh new volume. See also Relates https://github.com/elastic/cloud-on-k8s/issues/2806.
* We may want to migrate data away from the Pod before deleting its volume in the above situation.
* ECK could guess (best-effort) whether a Pod can be scheduled before triggering the rolling upgrade. It's probably very complicated to get right though.
* The user could indicate they require a "grow and shrink" approach to the change instead of a rolling upgrade. Where a new Pod is created with an empty volume, data gets migrated to it, and the old Pod and volume eventually get removed.

**Comment by barkbay** [comment]
I agree that this situation would be "exacerbated" by the use of local volume. But I like to see this kind of problem as a "data locality issue", where "locality" just means that the data is only available on a group of servers, and the Pod can't be scheduled anymore on that group.

For instance: Pods are using zonal persistent disks, there's no more resources available in this zone: the Pod can't be scheduled for the exact same reason. Very unlikely to happen on public clouds I guess, not sure on private ones...

**Comment by ravilr** [comment]
Pod priorities can be used to solve some of the concerns here.  Pods using local PVs are to be prioritized over other normal pods, as they can be scheduled only onto the node having the local volume. Kube Scheduler will evict lower priority pods to make room for a higher priority pod.

**Comment by shubhaat** [comment]
@mikeh-elastic I think this is broader than the particular usecase you are talking about.

**Comment by pebrc** [comment]
Change the label here as this is somewhat of a general K8s issue rather than a bug in ECK specifically
